src/mscorlib/src/System/ArraySegment.cs

```diff
@@ -106,12 +110,15 @@ public int Count
             }
         }
 
-        public Enumerator GetEnumerator()
+        public T this[int index]
         {
-            if (_array == null)
-                ThrowHelper.ThrowInvalidOperationException(ExceptionResource.InvalidOperation_NullArray);
-            Contract.EndContractBlock();
+            get { return _array[_offset + index]; }
```

Comment:
> Here is one place where this API could make code more readable. 

Agree. However, this example is not performance critical. The extra argument validation is not a problem.

> I don't think we want people manually writing as.Array[as.Offset + index] to get better perf

If you want to squeeze every last cycle out, you will probably write even something more convoluted. And that's perfectly fine. As I have said, the perf-sensitive algorithms should not be written against `ArraySegment<T>`.

I do not see a good reason why the ArraySegment indexer should deviate from framework design guidelines and not do argument validation.